cli/deno_dir.rs

```diff
@@ -203,298 +236,369 @@ impl DenoDir {
+    let source_code_headers = self.get_source_code_headers(&module_url);
+    // If source code headers says that it would redirect elsewhere,
+    // (meaning that the source file might not exist; only .headers.json is present)
+    // Abort reading attempts to the cached source file and and follow the redirect.
+    if let Some(redirect_to) = source_code_headers.redirect_to {
+      // E.g.
+      // module_name https://import-meta.now.sh/redirect.js
+      // filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/redirect.js
+      // redirect_to https://import-meta.now.sh/sub/final1.js
+      // real_filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/sub/final1.js
+      // real_module_name = https://import-meta.now.sh/sub/final1.js
+      let redirect_url = Url::parse(&redirect_to).expect("Should be valid URL");
+
+      let mut maybe_initial_module_url = maybe_initial_module_url;
+      // If this is the first redirect attempt,
+      // then maybe_initial_module_url should be None.
+      // In that case, use current module name as maybe_initial_module_url.
+      if maybe_initial_module_url.is_none() {
+        maybe_initial_module_url = Some(module_url.clone());
       }
+      // Recurse.
+      return self
+        .fetch_cached_remote_source(&redirect_url, maybe_initial_module_url);
     }
-  }
-
-  // If not remote file stop here!
-  if !is_module_remote {
-    debug!("not remote file stop here");
-    return Either::A(futures::future::err(
-      std::io::Error::new(
-        std::io::ErrorKind::NotFound,
-        format!("cannot find local file '{}'", filename),
-      ).into(),
-    ));
-  }
 
-  // If remote downloads are not allowed stop here!
-  if no_fetch {
-    debug!("remote file with no_fetch stop here");
-    return Either::A(futures::future::err(
-      std::io::Error::new(
-        std::io::ErrorKind::NotFound,
-        format!("cannot find remote file '{}' in cache", filename),
-      ).into(),
-    ));
+    // No redirect needed or end of redirects.
+    // We can try read the file
+    let filepath = self
+      .deps_cache
+      .location
+      .join(self.deps_cache.get_cache_filename(&module_url));
+    let source_code = match fs::read(filepath.clone()) {
+      Err(e) => {
+        if e.kind() == std::io::ErrorKind::NotFound {
+          return Ok(None);
+        } else {
+          return Err(e.into());
+        }
+      }
+      Ok(c) => c,
+    };
+    Ok(Some(SourceFile {
+      url: module_url.clone(),
+      redirect_source_url: maybe_initial_module_url,
+      filename: normalize_path(&filepath),
+      media_type: map_content_type(
+        &filepath,
+        source_code_headers.mime_type.as_ref().map(String::as_str),
+      ),
+      source_code,
+    }))
   }
 
-  debug!("is remote but didn't find module");
+  /// Asynchronously fetch remote source file specified by the URL following redirects.
+  fn fetch_remote_source_async(
```

Comment:
There's one quirk about this method regarding redirects: we only save `headers.json` file for requested URL and one for final URL after all redirects.

Consider such scenario (with `tools/http_server.py` running):
```
download http://localhost:4548/tests/001_hello.js
redirect to http://localhost:4546/tests/001_hello.js 
redirect to http://localhost:4545/tests/001_hello.js 
fetch from http://localhost:4545/tests/001_hello.js 
SourceFile(url: "http://localhost:4545/tests/001_hello.js", redirect_source_url: "http://localhost:4548/tests/001_hello.js")
cache headers for "http://localhost:4548/tests/001_hello.js": redirect -> http://localhost:4546/tests/001_hello.js
cache file "http://localhost:4546/tests/001_hello.js"
cache headers for "http://localhost:4546/tests/001_hello.js"
```

```
download http://localhost:4546/tests/001_hello.js
redirect to http://localhost:4545/tests/001_hello.js 
fetch from http://localhost:4545/tests/001_hello.js  // !!! unnecessary fetch, we already have cached file "http://localhost:4545/tests/001_hello.js"
SourceFile(url: "http://localhost:4545/tests/001_hello.js", redirect_source_url: "http://localhost:4546/tests/001_hello.js")
```

Should we save headers for all intermediate redirects 🤔 ? 

CC @ry

Reply:
> Should we save headers for all intermediate redirects 🤔 ?

Yes